### ISSUE TITLE
values vector in scale_color_manual

### ISSUE THREAD

**Issue by numpde** [tidyverse/ggplot2#4534]
The behavior of `scale_color_manual(values = ...)` seems to have changed recently and broke many of my plots. 

Consider this example:

```R
suppressPackageStartupMessages({
  library(ggplot2)
  library(dplyr)
})

color_map <- list(r = "red", g = "green", b = "blue")

randu %>%
  mutate(color_key = as.factor(if_else(x > y, "r", "g"))) %>%
  ggplot(aes(x = x, y = y, fill = color_key)) + 
  scale_fill_manual(values = color_map) +
  geom_point(shape = 21, size = 7)
```

It now produces a figure with the blue color _also_ in the legend. 
I find that the following works (blue does not appear in the legend):

```R
randu %>%
  mutate(color_key = as.factor(if_else(x > y, "r", "g"))) %>%
  {
    ggplot(., aes(x = x, y = y, fill = color_key)) + 
      scale_fill_manual(values = color_map, breaks = .$col) +
      # scale_fill_manual(values = color_map[.$col]) +  # Alternative
      geom_point(shape = 21, size = 7)
  }
```

But this code looks unhappy. Is there a more robust way, in particular without extra temporary variables?

Is there somethin like `breaks = .df$fill`? This would be useful in many contexts.

Sidenote. This would make more sense to me:
```R
scale_fill_manual(values = color_map)  # no blue in the legend
scale_fill_manual(values = color_map, breaks = names(color_map))  # with blue in the legend
```

==

For lols, this works, too:

```R
gglast <- function(what) {
  p <- ggplot2::last_plot()
  q <- p$mapping[[deparse(substitute(what))]]
  rlang::eval_tidy(q, data = p$data)
}

randu %>%
  mutate(color_key = as.factor(if_else(x > y, "r", "g"))) %>%
  ggplot(aes(x = x, y = y, fill = color_key)) + 
  scale_fill_manual(values = color_map, breaks = gglast(fill)) +
  geom_point(shape = 21, size = 7)
```

**Comment by teunbrand** [comment]
Issue #4511 also touches the extra colour key with predefined values.

**Comment by clauswilke** [comment]
Could you update to the latest ggplot2 on CRAN (3.3.5) and let us know if you still have this problem?

**Comment by numpde** [comment]
@clauswilke `sessionInfo()` shows ggplot2_3.3.5 in R 4.1.

Adding `limits = force` as proposed by @teunbrand in the linked issue also works (removes unused legend items).

**Comment by teunbrand** [comment]
It also has come to my attention that trying to do the opposite -displaying *fewer* legend keys- is also working in weird ways. In the code below, I would have expected both plots to do the same thing. Specifically, I would have expected the first plot to not have any grey points since the limits and values match up.

``` r
library(tidyverse)

p <- ggplot(mpg, aes(displ, hwy, colour = drv)) +
  geom_point(size = 2)

p +
  scale_colour_manual(
    values = c("dodgerblue", "tomato", "limegreen"),
    limits = c("4", "f", "r"),
    breaks = c("4", "r")
  )
```

![](https://i.imgur.com/poxAtRB.png)

``` r
p +
  scale_colour_manual(
    values = c("dodgerblue", "tomato", "limegreen"),
    limits = c("4", "f", "r"),
    breaks = function(x) c("4", "r")
  )
```

![](https://i.imgur.com/M0lO6Pp.png)

<sup>Created on 2021-09-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by yutannihilation** [comment]
Fixed by #4619
